lesson21/src/js/main.js

```diff
@@ -0,0 +1,212 @@
+import '../css/style.css'
+import * as loading from "./module/loading";
+import { createAttributedElements} from "./utils/createAttributedElements";
+
+const endPointURL = {
+    usersData: "https://myjson.dit.upm.es/api/bins/i55y"
+}
+
+const createErrorMessage = (error,element) => {
+    const errorMessage = createAttributedElements({
+        tag:"p",
+        valuesByAttributes:{
+            class:"error-message"
+        },
+        str:error
+    });
+    element.appendChild(errorMessage);
+}
+
+const fetchData = async (endpointURL,element) => {
+    const response = await fetch(endpointURL);
+    if(!response.ok){
+        console.error(`${response.status}:${response.statusText}`);
+        createErrorMessage("Communication with the server is broken.",element);
+        return;
+    }
+    const json = await response.json();
+    return json.data;
+}
+
+const fetchContentsData = (endPointURL,element,ms) => new Promise(resolve => setTimeout(() => resolve(fetchData(endPointURL,element)),ms));
+
+const table = document.getElementById("js-table");
+
+const initUsersData = async () => {
+    loading.showLoading(table);
+    try{
+        const usersData = await fetchContentsData(endPointURL.usersData,table,500);
+        if (!usersData.length) {
+            createErrorMessage("No user.",table);
+            return;
+        }
+        renderTableElements(usersData);
+        sortTableBody(usersData);
+    }catch(error){
+        console.error(error);
+    }finally{
+       loading.removeLoading(table);
+    }
+}
+
+const renderTableElements = usersData => {
+    renderTableHeader();
+    renderTableBody(usersData);
+}
+
+const usersTableColumn = {
+    "id"     : "ID",
+    "name"   : "名前",
+    "gender" : "性別",
+    "age"    : "年齢"
+}
+
+const renderTableHeader = () => {
+    const thead = document.createElement("thead");
+    const tr = document.createElement("tr");
+    thead.className = "bg-slate-500";
+
+    for (const [userKey, userValue] of Object.entries(usersTableColumn)) {
+        const th = createAttributedElements({
+            tag:"th",
+            valuesByAttributes:{
+                class:"text-sm text-white px-6 py-4",
+            },
+            str:userValue
+        });
+
+        tr.appendChild(th);
+
+        addSortButtons(userKey,userValue,th);
```

Comment:
Q.
Does the addSortButton function have to be done for every usersTableColumn?

Reply:
Only if the value in sortByUsersTableColumn matches the value in addSortButtons().
Is this not much better?

If addSortButtons() is used elsewhere, the
We need to get all elements of the TableHeader, which I think is a lot of processing, so we do it here.


````JavaScript
const addSortButtons = (userKey,userValue,th) => {
    for(const value of Object.values(sortByUsersTableColumn)){
        userValue === value && th.appendChild(createSortButtons(userKey));
    }
}
````

Reply:
Thank you for your comment!

I thought about your suggestion,
but this also uses a filter in the for statement,
and it might be better to separate it as an addSortButtons function like Maisan did, for better code visibility.

I hope you will find [CodeSandbox](https://codesandbox.io/s/maisannoke-ti-21-fork-lf8fwj) helpful.
And I would be glad to hear your opinion.

Reply:
CodeSandBoxまで用意していただいて、ありがとうございます！

はるさん、さえさんお二方のレビューをいただいて、
そもそもsortByUsersTableColumnをオブジェクトにしなくていいことに気づきました。

配列にすることで、コード自体がわかりやすくなるため、
あえてaddSortButtons関数として切り出すことはなさそうと思いました。

また、thを引数にする方法をご提案いただいたかと思いますが、すみません。。
もともと、thを引数にする形にしていたのですが、
appendChildする形がよいとのレビューをいただき、今の形になっております。

はるさんのコードをもとに、配列を使ったコードをさえさんにご提案いただいたので、修正してみました。
https://github.com/nakagawamai/js-lesson/commit/9d278ab27770b13493036ee07e1b3ba1359c8132

[Sae-san's Code.](https://github.com/nakagawamai/js-lesson/pull/23#discussion_r958974021)

````JavaScript
const renderTableHeader = () => {
 // omit

    //addSortButtons(userKey,userValue,th);
    if (sortCategories.includes(userKey)) {
      th.appendChild(createSortButtons(userKey));
    }
````

Reply:
@nakagawamai 
@sae-github 

> sortByUsersTableColumnをオブジェクトにしなくていい

本当ですね!
気づけませんでした...!
教えていただきありがとうございます。 
カテゴリーがsortCategoriesの配列に入っている時だけソートボタンがappendされ &
引数も少なくなってとても読みやすくなったと思います！
大変勉強になりました🙇‍♀️